File: _0_language/_concurrency/_3_synchronization_and_shared_resources/_synchronization/sync.py
import threading

# In Python, the threading module provides the Lock class to prevent multiple threads from accessing critical sections at the same time.

class Counter:
    def __init__(self):
        self.count = 0
        self.lock = threading.Lock()

    def increment(self):
        with self.lock:  # Acquire lock to ensure only one thread can increment at a time
            self.count += 1

    def get_count(self):
        return self.count

def increment_counter(counter):
    for _ in range(1000):
        counter.increment()

counter = Counter()

# Create two threads to increment the counter
thread1 = threading.Thread(target=increment_counter, args=(counter,))
thread2 = threading.Thread(target=increment_counter, args=(counter,))

thread1.start()
thread2.start()

thread1.join()
thread2.join()

print("Final count:", counter.get_count())  # Expected output: 2000


# We use a Lock to synchronize access to the increment() method, ensuring that only one thread can modify the count at a time.

File: _0_language/_concurrency/_3_synchronization_and_shared_resources/_1_synchronized_keyword/sync.go
package main

// In Go, you can use the sync.Mutex type to ensure that only one goroutine can access critical sections of code at a time.

import (
    "fmt"
    "sync"
)

type Counter struct {
    count int
    mu    sync.Mutex
}

func (c *Counter) increment() {
    c.mu.Lock()         // Lock the critical section
    c.count++
    c.mu.Unlock()       // Unlock the critical section
}

func main() {
    var counter Counter

    // Create two goroutines to increment the counter
    var wg sync.WaitGroup
    wg.Add(2)

    go func() {
        for i := 0; i < 1000; i++ {
            counter.increment()
        }
        wg.Done()
    }()

    go func() {
        for i := 0; i < 1000; i++ {
            counter.increment()
        }
        wg.Done()
    }()

    wg.Wait()  // Wait for both goroutines to finish
    fmt.Println("Final count:", counter.count)  // Expected output: 2000
}


// We use sync.Mutex to lock and unlock the critical section inside the increment() method, ensuring that only one goroutine can modify the counter at a time.
